### DISCUSSION TITLE
About kernel-headers

### DISCUSSION THREAD

**Discussion by skandbug** [yanivagman/BPFroid#10] [Q&A]
Hi，Thank you for giving us your help!

The bpf source code is not given in bpfroid, in
```python 
https://github.com/libbpf/libbpf/tree/d1fd50d475779f64805fdc28f912547b9e3dee8a
```
can be obtained successfully.
But KERN_HEADERS is not easy to get, as you mentioned
```python
KERN_HEADERS := /home/yaniv/src/thesis/pixel3a/android-kernel/headers/kernel-headers
```
I got it through
```python
https://github.com/Backup-Gits/android11-5.4
```

Use `make headers_install` to mention it, there are always some files missing, such as:` generated/autoconf, uapi/linux/ptrace.h`

It is possible to give a complete kernel-headers that you use when compiling, thank you.

**Comment by yanivagman** [comment]
Kernel headers are specific to the kernel you are using.
Even if I will provide the specific headers I used, they are useless if you are not running Android on a pixel 3a device.
So you will need to find a way to get all the headers for the kernel you are compiling to.

**Reply by skandbug** [reply]
Following the idea you provided, I generated the container in the host with your code. The version of my host is:
```python
5.15.0-46-generic.
```
I ran your program source code BPFroid in the container.

**Reply by skandbug** [reply]
I generated the headers with `make headers_install`, but it always suggests that some files are missing.

**Reply by skandbug** [reply]
I want to run the compiled BPFroid on Android, but I don't have the binary of BPFroid, so I need to compile it.
Does the compiled binary depend on the Android system version?

**Comment by yanivagman** [comment]
The headers that you need are of the **target** device, not of your host.
If your target is Android emulator - you can get the kernel headers from google site.
If your target is an Android device - you will need to get the headers from the device manufacturer.
Precompiled binary won't help you here - the code is compiled to specific kernel structs, and thus need to be compiled against the matching kernel headers

**Reply by skandbug** [reply]
Thank you for your prompt reply.

I am using the emulator and the version is: `5.4.47-01061-g22e35a1de440`.

You are an expert in this field, give google the download address of the header file if it is convenient.

**Reply by yanivagman** [reply]
This is the link for the kernel sources: https://source.android.com/docs/setup/build/building-kernels?hl=en
You will need to find the matching sources for the emulator version you used. It used to be called goldfish, but I see that they mention there that in Android 12 it is now changed, so I'm not sure about it

**Reply by skandbug** [reply]
Thank you for the resource, I will look for it along these lines first, thank you.

**Reply by skandbug** [reply]
I followed your instructions and completed the file related to `headers`, but I encountered multiple problems when compiling as follows

![image](https://user-images.githubusercontent.com/52226009/236462824-247a0791-21d9-492c-9329-78548a2d4262.png)

**Reply by skandbug** [reply]
How did you solve these problems during the compilation process?
Could you please advise me?

**Reply by JacquesJurado** [reply]
Maybe you can post your compilation steps of the kernel headers and of compilation process and repositories of the kernel and toolchain? I receive pretty similar errors e.g definition errors etc.. My assumption is that there is mismatch between the definitions of kernel headers and BPFroid. I did also like you cloned the toolchains of the repository "android-msm-bonito-4.9-android11"
![Screenshot from 2023-05-08 04-13-50](https://user-images.githubusercontent.com/30605546/236777058-8639aa41-3e06-4fe8-86f6-a1692921e5f6.png)

**Reply by yanivagman** [reply]
@JacquesJurado are you using an emulator or compiling for an Android device?

**Reply by JacquesJurado** [reply]
First I tried to build it for an goldfish kernel for later usage in an emulator. But this methodology didnt work cause of the same problems like @skandbug. Then i followed your advice for compiling it for a pixel 3A in the aarch64 docker container. But then there occured the compilation errors shown in the screenshot.

**Reply by yanivagman** [reply]
If you are using pixel 3a xl you can simply take the image included in the repository (the kernel headers are also there):
https://github.com/yanivagman/BPFroid/tree/main/pixel%203a

**Reply by skandbug** [reply]
Thanks, I'll try that later.

**Reply by skandbug** [reply]
The address you gave is a link, which shows all the source files of Android.
![image](https://user-images.githubusercontent.com/52226009/236868687-8c700a5c-5b9b-4d23-a9db-8b1cbd7914d2.png)

If it is based on the header file of Android11-5.4 (the Android kernel version is fixed), is it convenient for you to provide the header file or have you compiled an executable two-process program? We will be very grateful.

I seem to have encountered this problem too. @JacquesJurado

**Reply by yanivagman** [reply]
This link is only relevant for pixel 3a xl devices. I provided it since @JacquesJurado wrote that he is trying to compile for a pixel 3a device, for which it should work. To make BPFroid work on any other device/emulator you will have to provide the full kernel headers matching your platform (most probably you can download them from google).

Sorry that I can't help more with that, but this was a side project of mine and Android is not the main focus of my work these days...

**Comment by yanivagman** [comment]
Unfortunately, I don't remember having these errors

**Reply by yanivagman** [reply]
One thing I can't understand from your attached image - if you are compiling to an Android emulator, then why do you use the pixel3a folder? You should have your own kernel headers downloaded from google matching your emulator kernel version

**Reply by skandbug** [reply]
It's not compiling the emulator.
It is compiling bpfroid, the compilation process needs the kernel header files, I downloaded these files from GOOGLE, but during the compilation process there is a problem like the above picture.

**Reply by yanivagman** [reply]
Do you run the command as described in the readme?
`KERN_HEADERS=/path/to/android-goldfish-kernel make`
You should change the headers path to point to the sources you downloaded

**Reply by skandbug** [reply]
What I did was:
1. I downloaded the corresponding version of the header file here(`https://android.googlesource.com/platform/bionic/+/master/libc/kernel/`), but it prompted me that kconfig.h was missing when I compiled it.
2. I again in `https://android.googlesource.com/kernel/common/+/refs/heads/android11-5.4`
I extracted the source file and put it in the same level directory.
3. I modified the address of KERN_HEADERS in your source code to point to the downloaded header file directory.
After compiling, the error in the figure appears.
